### DISCUSSION TITLE
Likelihoods with shared components

### DISCUSSION THREAD

**Discussion by smar-git** [davidbolin/MetricGraph#8] [General]
Dear David and Alexandre
 
I have found a way to implement the mode where the data are not observations on a point ( as in your example) but averages over lines using the new A.local argument that Havard has added to the f() function.
We still do not manage to fit a joint model where we have two likelihoods with some shared components. This is the problem where you suggested to use the “group” option. But I still do not understand how to implement this. You saied that you were going to write some more documentation and I wanted to ask if you manage to do that.
Thanks a lot
Sara

**Comment by AlexandreBSimas** [comment]
Hi Sara,

Yes! Dont worry, I did not forget, I will do it. I wanted to have this done by now, however my kids are on vacation (requiring a lot of attention), and to make things worse my whole family was sick these past weeks, so did not manage to work much. 

In any case, I will update here when I implement it, will try to do it quickly. Also, very good news that you managed to implement it with this A.local argument! I was not aware of that argument, will take a look. I guess you can then probably implement it in inlabru. If you have any problems, I can try to help, you can message here. It is really super convenient to have these implementations in inlabru.

Best,
Alexandre

**Comment by smar-git** [comment]
Hi
Sorry, didn’t want to stress 😊
I’ll wait for your news then

From: Alexandre B. Simas ***@***.***>
Date: Thursday, 27 June 2024 at 17:54
To: davidbolin/MetricGraph ***@***.***>
Cc: Sara Martino ***@***.***>, Author ***@***.***>
Subject: Re: [davidbolin/MetricGraph] Likelihoods with shared components (Discussion #8)

Hi Sara,

Yes! Dont worry, I did not forget, I will do it. I wanted to have this done by now, however my kids are on vacation (requiring a lot of attention), and to make things worse my whole family was sick these past weeks, so did not manage to work much.

In any case, I will update here when I implement it, will try to do it quickly. Also, very good news that you managed to implement it with this A.local argument! I was not aware of that argument, will take a look. I guess you can then probably implement it in inlabru. If you have any problems, I can try to help, you can message here. It is really super convenient to have these implementations in inlabru.

Best,
Alexandre

**Reply by AlexandreBSimas** [reply]
Ahh, nono. Dont worry, you did not stress at all!
Sorry for the delay. I also expected to have this finished a while ago.

**Comment by AlexandreBSimas** [comment]
Hi Sara,

I have finished the vignette with the multiple likelihoods, with the example being exactly that model you mentioned. I used the group variable to handle the multiple likelihoods as I have mentioned to you before. I have modified the `graph_data_spde()` function to make it easier to handle these cases.

I have included examples with both INLA and inlabru.

The vignette is https://davidbolin.github.io/MetricGraph/articles/multi_likelihood.html

One thing is that since we are using the group variable to also handle the likelihoods, one might be confused on how to use replicates in such case, so I have also added examples in INLA and inlabru with multiple likelihoods and replicates.

Hope this helps you and solves your problem.

Best,
Alexandre

**Reply by AlexandreBSimas** [reply]
Ahh, just one thing. You will need to install the most recent version to be able to reproduce the vignette.

That is, you need to run:

```r
remotes::install_github("davidbolin/metricgraph", ref = "devel")
```

**Reply by karinalilleborge** [reply]
Hi Alexandre,
I am Sara's PhD student, and I have looked at the vignette you made for multiple likelihoods. Thank you very much for all the help so far!!

I have run through the code in your vignette, and I can reproduce the vignette results. :-) What I have trouble with is that if the two data.frames do not contain observations from the same set of locations I get an error when I run 
```
data_spde <- graph_data_spde(graph_spde = spde_model, 
                name = "field", likelihood_col = "likelihood",
                resp_col = "y",
                covariates = c("intercept_1", "intercept_2"))
```
The error message is 

> Error in ret_list[[lik]] <- ret : [[ ]] subscript out of bounds

I downloaded the .Rmd from github, and I have changed the cell where the 100 random locations are made (I don't necessarily have observations in the same locations):
```
obs_per_edge <- 100
obs_loc <- NULL
for(i in 1:(graph$nE)) {
  obs_loc <- rbind(obs_loc,
                   cbind(rep(i,obs_per_edge), 
                   runif(obs_per_edge)))
}
id1 <- seq(from=1, to=nrow(obs_loc), by=2)
id2 <- seq(from=2, to=nrow(obs_loc), by=2)
obs_loc1 <- obs_loc[id1,]
obs_loc2 <- obs_loc[id2,]
```
and I make observations as follows
```
y1 <- beta1 + u[id1] + sigma1.e * rnorm(n_obs/2)
y2 <- beta2 + u[id2] + sigma2.e * rnorm(n_obs/2)
```
And lastly I make data.frames for the two likelihoods:
```
df_graph1 <- data.frame(y = y1, intercept_1 = 1, intercept_2 = NA,
                        edge_number = obs_loc1[,1],
                        distance_on_edge = obs_loc1[,2],
                        likelihood = 1)
df_graph2 <- data.frame(y = y2, intercept_1 = NA,
                        intercept_2 = 1,
                        edge_number = obs_loc2[,1],
                        distance_on_edge = obs_loc2[,2],
                        likelihood = 2)      
df_graph <- rbind(df_graph1, df_graph2)  
```
All other cells are as in the vignette. I attach the modified .md here: 
[multi_likelihood.md](https://github.com/user-attachments/files/16310249/multi_likelihood.md)

I also tried to just remove one point from the second likelihood (so there are a total of 399 observations in the second likelihood, all in the same location as the first likelihood, but one location is not observed twice), and the same error is given. Am I doing something wrong when I change the data? 
I hope that my problem is clear from my examples here.

Best regards,
Karina Lilleborge

**Reply by AlexandreBSimas** [reply]
Hi Karina,

Thank you for your message. It really was supposed to work in general, not only when having observations on the same locations. 

I have pushed a bugfix. 

Please, could you try again and confirm whether it fixed for you? 

Also, for you to get good estimates on the measurement error in the modified vignette you sent (since you were dividing the number of observations by 2), you probably will need to increase `obs_per_edge` to 150 or 200. 

Best,
Alexandre

**Reply by karinalilleborge** [reply]
Hi Alexandre,
I did the remotes::install_github()- which did update a few things, but I still get the same error when I run the code from the start. I am currently trying the last example where I have removed only one observation in the df_graph2:
```
df_graph <- rbind(df_graph1, df_graph2[-1,])        
```
Otherwise, my code is the same as the vignette.
Thank you for the very quick response!
Bests,
Karina

**Reply by AlexandreBSimas** [reply]
Weird, here it is working just fine.

Did you close R, and reopened it (to load the new version of metricgraph)?

**Reply by karinalilleborge** [reply]
Sorry, I just cleared the workspace, but didn't load the new package, my mistake.
Thank you for fixing it so quickly! It seems to work now. 😊 
Bests,
Karina

**Reply by AlexandreBSimas** [reply]
Great that it is working! My pleasure!

If you run into any other issues, just let me know.

**Comment by smar-git** [comment]
﻿Thanks a lot!
I’ll have a look at the vignette
sara

From: Alexandre B. Simas ***@***.***>
Date: Sunday, 7 July 2024 at 03:09
To: davidbolin/MetricGraph ***@***.***>
Cc: Sara Martino ***@***.***>, Author ***@***.***>
Subject: Re: [davidbolin/MetricGraph] Likelihoods with shared components (Discussion #8)

Ahh, just one thing. You will need to install the most recent version to be able to reproduce the vignette.

That is, you need to run:

remotes::install_github("davidbolin/metricgraph", ref = "devel")
